chore: add notification settings section

astone123 opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes https://github.com/cypress-io/cypress/issues/26778

### Additional details
<!-- Examples:
- Why was this change necessary?
- What is affected by this change?
- Any implementation details to explain?
-->

This PR adds a section to device settings called "Desktop notifications" that will allow users to configure which notifications they see. 

**NOTE** that these configuration values aren't used for anything yet. They will be used in https://github.com/cypress-io/cypress/issues/26788 to decide when to send notifications.

See the related issue for ClickUp requirements and Figma mockups

### Steps to test
<!--
For non-trivial behavior changes, list the steps that a reviewer should follow to validate the new behavior.
This is not meant to be the only testing performed by a reviewer, just the "happy path" that leads to the new behavior.
-->

1. Launch Cypress
2. Navigate to settings (gear icon in left nav bar)
3. Open device settings
4. Scroll down to Desktop notifications section
5. Click values in the form and verify that values are updated correctly
6. Refresh/reload the app and ensure that your selections were persisted

### How has the user experience changed?
<!-- Provide before and after examples of the change.
Screenshots or GIFs are preferred. -->

<img width="1187" alt="Screenshot 2023-05-30 at 4 12 31 PM" src="https://github.com/cypress-io/cypress/assets/14275198/dd30859a-842c-498d-a50d-939fa1af04f6">

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

astone123 committed:
feat: enable notifications banner

astone123 committed:
fix issue where specs list container would scroll when banner was on screen

astone123 committed:
remove subscription and emitter

astone123 committed:
revert unnecessary changes

warrensplayer committed:
move specs list to use grid layout (#26881)

Co-authored-by: astone123 <adams@cypress.io>

astone123 committed:
chore: add notification settings section

astone123 commented on packages/data-context/src/actions/LocalSettingsActions.ts:
```diff
@@ -11,25 +11,34 @@ export interface LocalSettingsApiShape {
 }
 
+// Combine two objects into one either using Lodash merge strategy or the spread operator
+function combine (first: Object = {}, second: Object = {}, merge: boolean) {
```
I added this because I needed to overwrite the `notifyWhenRunCompletes` property, but when we use `_.merge`, Lodash merges the list items together instead of overwriting the property.

mike-plummer replied:
Could we use [`_.mergeWith`](https://lodash.com/docs/4.17.15#mergeWith) and provide a customizer to handle the array case? The non-merge case loses a lot of the recursive behavior we get with `merge`

astone123 commented on packages/graphql/src/schemaTypes/objectTypes/gql-LocalSettings.ts:
```diff
@@ -36,6 +36,15 @@ export const LocalSettingsPreferences = objectType({
     t.dateTime('dismissNotificationBannerUntil')
+    t.boolean('notifyWhenRunStarts')
+    t.boolean('notifyWhenRunStartsFailing')
+    t.json('notifyWhenRunCompletes', {
```
I tried to use a GraphQL List type with an enum for this, but I couldn't get the initial values to load properly unless I did this.

astone123 commented on packages/app/src/settings/device/NotificationSettings.vue:
```diff
@@ -0,0 +1,154 @@
+          keypath="settingsPage.notifications.notReceivingNotifications"
+          class="pt-[15px] text-gray-600 text-sm font-normal"
+        >
+          <ExternalLink href="#">
```
We will add the link to the docs for this after https://github.com/cypress-io/cypress-documentation/issues/5280 is completed

mike-plummer replied:
Maybe better to pull out a `computed` to generate the `getUrlWithParams` with a `#` base URL so we can at least put the `TODO` alongside the needed work. Things like this inside the template tend to get lost in the noise

astone123 committed:
fix component test

mike-plummer commented on packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts:
```diff
@@ -439,9 +439,10 @@ export const mutation = mutationType({
         type: nonNull(arg({
           type: PreferencesTypeEnum,
         })),
+        merge: booleanArg({ description: 'Whether to merge new preferences or override them' }),
```
Might be a bit clearer to specify a `default` value here

astone123 replied:
Ended up getting rid of this because of your suggestion to use `_.mergeWith`

mike-plummer commented on packages/frontend-shared/src/components/Checkbox.vue:
```diff
@@ -38,19 +38,23 @@
-const emit = defineEmits<{
-  (event: 'update:modelValue', value: boolean): void
+const emits = defineEmits<{
+  (event: 'update:modelValue'): void
```
This does't need to emit the updated modelValue?

astone123 replied:
I don't think so? 🤔 There's some Vue magic going on here that I don't quite understand

mike-plummer commented on packages/frontend-shared/src/components/Checkbox.cy.tsx:
```diff
@@ -1,16 +1,29 @@
-    >
-      <span class="text-gray-800 font-light">Show the welcome guide when opening Cypress.</span>
-    </Checkbox>))
+    cy.mount({
+      components: { Checkbox },
+      render () {
+        return h(Checkbox, {
+          label: 'Welcome guide settings',
+          id: 'welcome-opt-out',
+          modelValue: reference.value,
+          'onUpdate:modelValue': updateStub,
+        })
+      },
+    })
```
Why do we have to use `h` syntax here? Am I missing something that doesn't work with standard JSX?

astone123 replied:
I'm trying to remember why I did this 🤔 will try just using jsx

astone123 replied:
Updated this to use jsx

mike-plummer commented on packages/app/src/settings/device/NotificationSettings.vue:
```diff
@@ -0,0 +1,154 @@
+  })
+}
+
+watch(() => listRef.value, async (newVal) => {
```
Maybe a good use case for [`watchDebounced`](https://vueuse.org/shared/watchDebounced/) to minimize the number of writes

astone123 replied:
Good idea, added

mike-plummer commented on packages/app/src/settings/device/NotificationSettings.vue:
```diff
@@ -0,0 +1,154 @@
+import Switch from '@packages/frontend-shared/src/components/Switch.vue'
+import SettingsSection from '../SettingsSection.vue'
+import { NotificationSettingsFragment, SetNotificationSettingsDocument, NotificationSettings_ShowNotificationDocument } from '../../generated/graphql'
+import Checkbox from '@cy/components/Checkbox.vue'
```
FYI - there is a [`Checkbox` available from the design system](https://design.cypress.io/components/vue/Checkbox.html). Is the intent to use our existing implementation?

Edit: DS version is restricted to `boolean` so maybe not a real option without some serious updates, but figured I should still ask the question

astone123 replied:
I didn't realize that there was one... I don't think it fits this use case easily, so for now we should move forward with this I think. It looks like this checkbox component wasn't used anywhere else except its test before this

astone123 committed:
feedback

jordanpowell88 commented on packages/frontend-shared/src/components/Checkbox.cy.tsx:
```diff
@@ -2,15 +2,18 @@ import { ref } from 'vue'
-    >
-      <span class="text-gray-800 font-light">Show the welcome guide when opening Cypress.</span>
-    </Checkbox>))
+    cy.mount(<Checkbox label='Welcome guide settings' id="welcome-opt-out" modelValue={reference.value} onUpdate:modelValue={updateStub} />)
```
`refrence.value` reads much better than `value.value` 👍

jordanpowell88 commented on packages/frontend-shared/src/components/Checkbox.vue:
```diff
@@ -38,19 +38,23 @@
+const props = withDefaults(defineProps<{
   id: string
-  modelValue: boolean
+  modelValue: any
```
Why do we need to change the type here?

astone123 replied:
Because in this case the value that I'm passing is a list of strings, not a boolean

jordanpowell88 commented on packages/frontend-shared/src/locales/en-US.json:
```diff
@@ -480,6 +480,22 @@
+      "failed": "Failed",
+      "canceled": "Canceled",
+      "errored": "Errored",
+      "sendTestNotification": "Send a test notification",
```
Is this for the `Troubleshoot` link?

astone123 replied:
This is for the "Send a test notification" button in the settings section

jordanpowell88 reviewed (approved):
LGTM. Left a few small questions but overall great job

astone123 committed:
use debounce

warrensplayer commented on packages/app/src/settings/SettingsContainer.vue:
```diff
@@ -23,6 +23,7 @@
       >
         <ExternalEditorSettings :gql="props.gql" />
         <ProxySettings :gql="props.gql" />
+        <NotificationSettings :gql="props.gql" />
```
I think the max-height of the SettingsCard this is inside of should be increased so it does not scroll by default.

warrensplayer reviewed (approved):
Looking pretty good. Going to approve, but one thing to look at is the default settings not getting set correctly.   I see the following:

<img width="942" alt="image" src="https://github.com/cypress-io/cypress/assets/1702361/6d8f9c2a-19f4-42db-aa04-1d9d9c0f2d5c">

I do not have a default setting for "Notify when run completes" with the "failed" checked.

astone123 committed:
fix notifyWhenRunCompletes default

astone123 commented:
> Looking pretty good. Going to approve, but one thing to look at is the default settings not getting set correctly. I see the following:
> 
> <img alt="image" width="942" src="https://user-images.githubusercontent.com/1702361/242975953-6d8f9c2a-19f4-42db-aa04-1d9d9c0f2d5c.png">
> I do not have a default setting for "Notify when run completes" with the "failed" checked.

@warrensplayer I fixed this in https://github.com/cypress-io/cypress/pull/26885/commits/55eb7f3abde308f768b013a0fd4c127e771b3243. In the resolver for `notifyWhenRunCompletes`, I was calling 
```js
ctx._apis.localSettingsApi.getPreferences()
```
which doesn't take into account any defaults that we've gathered and reads directly from the state file. What I actually want is
```js
ctx.coreData.localSettings.preferences.notifyWhenRunCompletes
```

The reason that it was working in the e2e tests is because [I was mocking that function](https://github.com/cypress-io/cypress/pull/26885/commits/55eb7f3abde308f768b013a0fd4c127e771b3243#diff-89d4c26d4b05ec49dcd7f071241d53d24e791ba66e6aa229a5710091dd79c99eL400) to return that preference. I removed the mock now so that we are actually testing whether or not the GQL resolver gets the default like it should
